Task_1_2

Sharapov-Maksim opened the pull request.

Sharapov-Maksim committed:
First version of task

Completed task with no test files at this moment

Sharapov-Maksim committed:
General revision

Added tests, fixed bugs, public contract revised.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+        char substr[] = new char[substrIn.length()];
+        substr = substrIn.toCharArray();
+        int lenSubStr = substr.length;
+        BufferedReader readerq = null;
+        FileInputStream reader = null;
+        long pos = 0; //position of start substring entry
+        //TODO два буфера, каждый размером в подстроку, конкат, и поиск подстроки в новой
+        try {
+            readerq = new BufferedReader(new InputStreamReader(new FileInputStream(fname), StandardCharsets.UTF_8));
+            BufferedReader in = new BufferedReader(new InputStreamReader(new FileInputStream(fname), "UTF-8"));
+            char[] strForSearch = new char[lenSubStr * 2];
```
Не все классы, реализующие AutoCloseable, закрываются, либо закрываются корректно.
Для того, чтобы понять, реализует класс этот интерфейс, или нет, можно найти его на docs.oracle.com и в начале страницы документации посмотреть на All Implemented Interfaces.

vlitomsk replied:
try-with-resources в помощь. Про это я писал в письме.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+                resList.add(0L);
+                return resList;
+            }
+            while (l2 != -1) {
+                strForSearch = bufcat(buffer1, buffer2);
+
+                /*Implemented z-function algorithm to find all substrings in string strForSearch
+                * works for linear time O(N+M)*/
+
+                char[] str = new char[lenSubStr * 3];
+                str = bufcat(substr, strForSearch);
+                int zArr[] = new int[str.length];
```
Выделение памяти в куче (в том числе и вызовом bufcat) -- дорогая операция. 
Предлагаю избавиться от лишних аллокаций.

vlitomsk replied:
Также предлагаю сделать сплошной буфер одним участком памяти, а не два отдельных. Таким образом можно избавиться от bufcat.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+                l2 = in.read(buffer2, 0, lenSubStr);
+            }
+        } catch (IOException e) {
+            System.out.println("IO Error: " + e);
```
Что, если пользователь в случае любых ошибок хочет завершать исполнение программы?
Предлагаю оставить обрабтку исключений пользователю этой функции.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+     * @return ArrayList with all entrуes of substrIn in file
+     */
+
+    public static ArrayList<Long> substrSearchInFile (String fname, String substrIn) {
```
Предлагаю для универсальности функции сделать так, чтобы она принимала не путь до файла, а просто поток ввода. Таким образом можно будет подавать её не только файлы, но и строки, трубы, и вообще все что вписывается в концепцию потока ввода.

vlitomsk commented on task_1_2/src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Test
+    public void substrSearchInFile() throws IOException {
+        // test 1
```
Предлагаю разбить это на отдельные тест-кейсы.
А то приходится думать, связан ли код двумя экранами ниже с тем, что двумя экранами выше.
Да хотя бы тесты внутри этой функции проверяют разное поведение целевой функции, например, 1. строчка найдена 2. строчка не найдена.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+            char buffer2[] = new char[lenSubStr];
+            int l1 = in.read(buffer1, 0, lenSubStr);
+            if (l1 < lenSubStr) {
+                return resList;
```
Предлагаю эти два примера добавить в тестовый класс.
substr: abc
1. str: abcaabc
2. str: abcabcaabc

vlitomsk replied:
смысл в том, что последние считанные символы могут быть суффиксом искомой строки.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+        long pos = 0; //position of start substring entry
+        //TODO два буфера, каждый размером в подстроку, конкат, и поиск подстроки в новой
+        try {
+            readerq = new BufferedReader(new InputStreamReader(new FileInputStream(fname), StandardCharsets.UTF_8));
```
это не используется. надо удалить.

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+
+                l1 = l2;
+                pos -= l2;
+                for (int i = 0; i < lenSubStr; i++) {
+                    buffer1[i] = buffer2[i];
+                }
```
Заменить на более эффективную Sytem.arraycopy

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,115 @@
+        return resList;
+    }
+
+        private static char[] bufcat (char[] b1, char[] b2){
+        char[] res = new char[b1.length+b2.length];
+        int l1 = b1.length;
+        int l2 = b2.length;
+        for(int i = 0; i<l1; i++){
+            res[i] = b1[i];
+        }
+        for(int i = 0; i<l2; i++){
+            res[l1+i] = b2[i];
+        }
+        return res;
+    }
+}
```
это после исправления замечаний не понадобится.

vlitomsk commented on task_1_2/src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,144 @@
+        }
+
+        // test 4
+        String fname4 = "test4.txt"; // put your file name here
+        substrUTF_8 = new String("Монада - это моноид в категории эндофункторов".getBytes(), "UTF-8");
+        BufferedWriter out4 = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(fname4), StandardCharsets.UTF_8));
```
А если в файле test4.txt лежат какие-то нужные данные? Может быть, его создал кто-то до нас?
Для таких целей обычно используются временные файлы. См. File.createTempFile

vlitomsk commented on task_1_2/src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,144 @@
+        //test 5
+        String fname5 = "test5.txt"; // put your file name here
+        substrUTF_8 = new String("пирог".getBytes(), "UTF-8");
+        BufferedWriter out5 = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(fname5), StandardCharsets.UTF_8));
```
try-with-resources.
и в других местах в этой функции тоже.

vlitomsk reviewed (changes requested):
Замечания нужно исправить.

Sharapov-Maksim committed:
Delete notes4.txt

Sharapov-Maksim committed:
Delete input.txt

Sharapov-Maksim committed:
Fixes

Внесены исправления на основе ревью

vlitomsk commented on task_1_2/src/main/java/Main.java:
```diff
@@ -0,0 +1,83 @@
+                l1 = l2;
+                pos -= l2;
+                System.arraycopy(str, lenSubStr*2, str, lenSubStr, lenSubStr);
+                l2 = in.read(str, lenSubStr*2, lenSubStr);
```
А если под конец потока будет прочитано меньше байт, чем запрошено? (длина потока не кратна длине подстроки)

Sharapov-Maksim replied:
Если будет считано 0 байт, то встречен конец строки и в l2 вернётся -1. Если же считано меньше, чем запрошено, то символы, идущие после этих символов не изменятся, и произойдёт БАН. Сейчас поправлю)

vlitomsk commented on task_1_2/src/test/java/Search0Test.java:
```diff
@@ -0,0 +1,31 @@
+public class Search0Test {
+
+    @Test
+    public void substrSearchInStream() throws IOException {
```
можно было не создавать несколько классов под тест одной функции.
в одном классе может содержаться несколько @Test -методов.

vlitomsk replied:
ну да ладно уже..

vlitomsk reviewed (changes requested):
Вопрос по l2 = in.read(...)

Sharapov-Maksim committed:
Fix

Исправлен баг,  когда из-за считывания неполного полубуфера (когда l2 < lenSubStr) можно было получить лишние ответы.

vlitomsk reviewed (changes requested):
Ок, принято!
